.github/workflows/main.yml

```diff
@@ -19,25 +19,77 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      # TODO: cache all the setup
-
       - name: Setup directories
         run: |
           # create directory that will be on the PATH
           mkdir -p ~/.local/bin
           echo "${HOME}/.local/bin" >> $GITHUB_PATH
           mkdir -p ~/source
+     
+      - uses: actions/cache@v2
+        id: cache-gcc
+        with:
+          path: ~/.local/bin/gcc
+          key: ${{ runner.os }}-gcc
+
+      - uses: actions/cache@v2
+        id: cache-clang
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-clang
+      
+      - uses: actions/cache@v2
+        id: cache-clang-format
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-clang-format
+
+      - uses: actions/cache@v2
+        id: cache-libc6-i386
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-libc6-i386
+
+      - uses: actions/cache@v2
+        id: cache-linux-module
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-linux-module
+
+      - uses: actions/cache@v2
+        id: cache-make
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-make
+
+      - uses: actions/cache@v2
+        id: cache-pip
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
 
       - name: Install gcc, clang, clang-format
         run: |
           sudo add-apt-repository ppa:ubuntu-toolchain-r/test
           sudo apt-get update
-          sudo apt-fast -y install gcc-6
-          sudo apt-fast -y install clang-5.0
-          sudo apt-fast -y install clang-format-5.0
-          sudo apt-fast -y install libc6-i386
-          # for vcan module
-          sudo apt-fast -y install linux-modules-extra-$(uname -r)
+        
+        if: steps.cache-gcc.outputs.cache-hit != 'true'
+        run: sudo apt-fast -y install gcc-6
+
+        if: steps.cache-clang.outputs.cache-hit != 'true'
+        run: sudo apt-fast -y install clang-5.0
+
+        if: steps.cache-clang-format.outputs.cache-hit != 'true'
+        run: sudo apt-fast -y install clang-format-5.0
+
+        if: steps.cache-libc6-i386.outputs.cache-hit != 'true'
+        run: sudo apt-fast -y install libc6-i386
+
+        # for vcan module
+        if: steps.cache-linux-module.outputs.cache-hit != 'true'
+        run: sudo apt-fast -y install linux-modules-extra-$(uname -r)
+
+        run: |
```

Comment:
I'm fairly certain you can only have one `if` and one `run` per step - this is a limitation of the YAML format. So you'll need to split these into multiple steps, one per if/run.